### ISSUE TITLE
Windows AAD login

### ISSUE THREAD

**Issue by ChoOo7** [paolosalvatori/ServiceBusExplorer#607]
Hello

Can we have an Azure Active directory login to have access to our services bus ?

We don't want to have any password / keys on our computers

Thanks

**Comment by ErikMogensen** [comment]
@ChoOo7, that's something that's on the roadmap. However, currently 95% of the code Service Bus Explorer is using to communicate with queues and topics is using the oldest Service Bus SDK. The remaining 5% is using the latest Service Bus SDK. For several reasons we want to migrate everything to the latest SDK and that has a higher priority for me. 

However, you or anyone else is most welcome to come up with a design suggestion for AAD login. I recommend starting with that before creating a PR since this a complicated issue.

**Comment by Rookian** [comment]
I think the easiest solution is to use `AzureCliCredential` or `InteractiveBrowserCredential` when authenticating against Azure Service Bus. The new SDK already provides an option to use `TokenCredential`.
```csharp
 public ServiceBusClient(string fullyQualifiedNamespace, TokenCredential credential);
 ```

**Comment by SeanFeldman** [comment]
@Rookian 
>  However, currently 95% of the code Service Bus Explorer is using to communicate with queues and topics is using the oldest Service Bus SDK.

**Comment by kensykora** [comment]
I really want this feature so we can stop using access keys!

**Comment by jwerner77** [comment]
Azure AD auth will be a huge win for this tool!  There is some very valuable functionality such as fix/resubmit from dead letter queue which are not available in the portal.  However the only way for a team to use this is they all have to have connection string with the keys which is not secure.

**Comment by rangp** [comment]
Is there any update on this? 
I just looked at the code and it seems as if the old sdk has been removed. With the new sdk it should be fairly simple to integrate this, since both the ServiceBusClient and ServiceBusAdministrationClient support passing a TokenCredential. 

I can see two ways of integrating this:

1. Add a ui feature that allows one to specify whether to authenticate via active directory or not. Something like this. And if specified, a flag is passed that allows the ServiceBusClient and ServiceBusAdministrationClient to be constructed with InteractiveBrowserCredentials. This would then open a browser window where one can sign in and the rest should be handled by the sdk out of the box
![mockup](https://user-images.githubusercontent.com/8915141/165271679-a2584d06-70e8-431d-b96c-dd2bad675c75.png)

2. add a magic connecting string. This is how the [sqlclient](https://docs.microsoft.com/en-us/sql/connect/ado-net/sql/azure-active-directory-authentication?view=sql-server-ver15) handles it too. One could for example specify a connection string like this

`Endpoint=sb://....servicebus.windows.net/;Authentication=Active Directory Interactive`

Then provide a factory for creating a ServiceBusClient and ServiceBusAdministrationClient that can interpret the connection string and constructs the client with InteractiveBrowserCredentials. Use the factory instead of constructing the clients directly. This solution is not as elegant as the first, but I think it should be very simple to implement, as no ui changes are needed and the code only needs to be changed in a handful of places.

**Comment by diligent176** [comment]
This would be great to have as we move over to RBAC model for service bus.

**Comment by bjoljo** [comment]
Azure AD login would be very usefull, please let us have the option to move away from connection strings.

**Comment by malsey** [comment]
would be great to have this supported

**Comment by brswanson** [comment]
Would also really like to have AD auth for this 🙏

**Comment by madmalkav** [comment]
Would be incredibly useful to have this implemented.

**Comment by zyofeng** [comment]
I would say this is an essential feature as our company moves to disable local authentication (SAS key).
Until then, the Azure portal provides an AAD integrated service bus explorer that one can use.

**Comment by sovaska** [comment]
Pretty much stopped using Service Bus Explorer since Managed Identity is best practice for accessing Service Bus. Would love to still use it, but without Managed Identity support it’s not possible.

**Comment by SeanFeldman** [comment]
@jooooel, unfortunately, the code hasn’t been migrated yet.

**Comment by mennolaan** [comment]
Also looking forward to be able to use managed identities. Many companies now move away from the sas tokens and only allow managed identity, thus removing the possibility to servicebus explorer.

**Comment by Merlijnv** [comment]
@SeanFeldman is there a way to support the transition? idk how much work is left in migrating?

**Comment by mennolaan** [comment]
> @SeanFeldman is there a way to support the transition? idk how much work is left in migrating?

Yes i second this, would not mind contributing to this.

**Comment by SeanFeldman** [comment]
> is there a way to support the transition? idk how much work is left in migrating?

@Merlijnv, absolutely. The biggest work is in transitioning off the old SDK to the new multiple SDKs (ASB, EH). This is the first step towards supporting AAD.

> Yes i second this, would not mind contributing to this.

@mennolaan, that's terrific! I've created a [designated milestone](https://github.com/paolosalvatori/ServiceBusExplorer/milestone/29) to capture the work needed to be done to migrate SDKs.
Have a look at the issues, feel free to add additional ones that are relevant for that effort.

cc @paolosalvatori, @ErikMogensen

**Comment by Merlijnv** [comment]
@SeanFeldman what is the best starting point? 
Is there a discord or something to chat?
I have time tomorrow and the day after tomorrow 😄

**Comment by SeanFeldman** [comment]
@Merlijnv, that's the tricky part. We're all fully dispersed with availability constraints.
If I had to do that, I'd get familiar with the codebase first and then try to see if replacing certain code paths with the newer SDKs is a go or not. We've tried that with `ServiceBusHelper` by introducing [`ServiceBusHelper2`](https://github.com/paolosalvatori/ServiceBusExplorer/blob/main/src/ServiceBus/Helpers/ServiceBusHelper2.cs), as an experiment.

**Comment by Merlijnv** [comment]
@SeanFeldman
I see some other issue #527 saying that the biggest blocker for .net 5 is WindowsAzure.ServiceBus should that be my first focus?

Becides that I see also #286 talking about cross platform idk if that's something you guys are still looking into?

To check all WindowsAzure.ServiceBus, Microsoft.ServiceBus.Messaging and such should be migrated to [Azure.Messaging.ServiceBus](https://www.nuget.org/packages/Azure.Messaging.ServiceBus/). Is that correct?

**Comment by ErikMogensen** [comment]
@Merlijnv, WindowsAzure.ServiceBus is the main blocker so if you got rid of it, it would be wonderful. It is a huge job though. You have to migrate to the different SDKs that support ServiceBus, Event Hubs and Relay. There are features missing so keeping backwards compatibility may not be possible. I suggest you choose one of the services to start with. I know that the latest SDK for Event Hubs management is lacking support for authorization using SAS so you have to add a login for Entra authentication for that. 

We decided that cross platform will have to wait at least until WindowsAzure.ServiceBus has been removed.

**Comment by Merlijnv** [comment]
@ErikMogensen Ah okay thanks for the explanation. I will prob begin with servicebus because at our project we don't use the other 2 so I have no way to test. Prob will create a way to test when I am done with servicebus and nobody else started on it. I will try and expand the servicebushelper2 with all the methods that can be used in the new sdk. Is there anything else I should know or pay attention to?

**Comment by ErikMogensen** [comment]
Sounds good. There is a guide at https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/MigrationGuide_WindowsAzureServiceBus.md

Any feature that is missing in Azure.Messaging.ServiceBus and not documented as missing is interesting so please put a comment here. @paolosalvatori can notify the Service Bus Product Group about it and there is a chance that they may implement it.

There is a large PR for Event Grid, #737, that affects MainForm.cs and perhaps other central parts so please keep an eye on that one to avoid merge issues.

**Comment by SeanFeldman** [comment]
@Merlijnv, everything that Erik said.
I can help with the ASB, assuming I have availability to do so. If you want to start a PR and ping me with some questions as you progress, we can try that. 

> I will try and expand the servicebushelper2 with all the methods that can be used in the new sdk.

`[ServiceBusHelper2](https://github.com/paolosalvatori/ServiceBusExplorer/blob/main/src/ServiceBus/Helpers/ServiceBusHelper2.cs) is just a helper. Don't worry about it too much. If needs to go, it can be removed as well.

Regarding EventGrid PR - I'm looking after that one. We don't need to worry about it until it's merged into `master` and then whatever you'll have at that point would need to be rebased. But let's not worry about it ahead of the time.

**Comment by 0xced** [comment]
> We decided that cross platform will have to wait at least until WindowsAzure.ServiceBus has been removed.

Are you implying there's a plan to rewrite ServiceBusExplorer with [Avalonia UI](https://avaloniaui.net/) or [Uno Platform](https://platform.uno/)? Because cross platform WinForms is definitely [not](https://github.com/dotnet/winforms/issues/8828#issuecomment-1468967026) [happening](https://github.com/mono/mono/issues/6701#issuecomment-1619992497).

**Comment by Merlijnv** [comment]
@SeanFeldman Do you maybe have some time maybe next monday at 10am (that's 18.00 my time)? I need some quick walkthrough as to which code will be hit first and how everything gets called.

**Comment by Merlijnv** [comment]
@paolosalvatori so far it seems like getting queue message counts is missing or I can't find it. QueueDescription has been replaced with QueueProperties but that does not include MessageCountDetails.

**Comment by SeanFeldman** [comment]
> @SeanFeldman Do you maybe have some time maybe next monday at 10am (that's 18.00 my time)? I need some quick walkthrough as to which code will be hit first and how everything gets called.

Sorry, can't make it in the AM hours. You might have more luck with @paolosalvatori or @ErikMogensen as they are in EU time zones. Queue-related forms and controls might be the simplest path. Topics/subscriptions are a duplication. Something to be aware of, BTW.

**Comment by Merlijnv** [comment]
@SeanFeldman @paolosalvatori @ErikMogensen is the retry helper still needed when using servicebusadministrationclient?

**Comment by SeanFeldman** [comment]
`ServiceBusAdministrationClient` still can throw [exceptions](https://learn.microsoft.com/en-us/dotnet/api/azure.messaging.servicebus.administration.servicebusadministrationclient.createqueueasync?view=azure-dotnet#azure-messaging-servicebus-administration-servicebusadministrationclient-createqueueasync(azure-messaging-servicebus-administration-createqueueoptions-system-threading-cancellationtoken)) that would need to be logged and some retried.

@Merlijnv, I suggest you open a PR (even if it's only at a starting point) and have this kind of Q&A on the PR rather than the issue. The issue is more suited to the problem description, status updates, and information not directly implemented. PR is more suited for the discussions and decision-making related to the changes in that PR. Would you agree?

**Comment by ancientpanda** [comment]
Any update on this? Entra/AAD support would make this tooling very ideal over having to use less secure SAS tokens.

**Comment by maztan** [comment]
Isn't Entra ID Authentication officially supported for Service Bus https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-authentication-and-authorization

**Comment by ErikMogensen** [comment]
> Isn't Entra ID Authentication officially supported for Service Bus https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-authentication-and-authorization

Yes, it is. Why do you ask? The issue here is that it is a lot of work to add it especially since the code uses two different SDKs and one of them is deprecated.

**Comment by maztan** [comment]
> > Isn't Entra ID Authentication officially supported for Service Bus https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-authentication-and-authorization
> 
> Yes, it is. Why do you ask? The issue here is that it is a lot of work to add it especially since the code uses two different SDKs and one of them is deprecated.

So it all comes down to migrating the app to the latest SDK

**Comment by SeanFeldman** [comment]
Yes. Sounds trivial but it's not.

**Comment by maztan** [comment]
Ok, will give it a shot after I read all the documentation and understand the changes made to the SDK

**Comment by maztan** [comment]
I am a little confused. There seems to be an option to authenticate using Entra ID in the current version of the application but only for Azure Event Grid and the other option that uses SAS for Service Bus. Is this because the Entra ID is only partially integrated in current version?

**Comment by SeanFeldman** [comment]
That's because EventGrid was implemented recently as an independent capability. It's also using a different SDK, not ASB SDK.

**Comment by maztan** [comment]
I think it would be easier to create new projects that use new SDK version and move functionality over from the current ones as there will be too much mess otherwise. Just my opinion.

**Comment by SeanFeldman** [comment]
https://github.com/paolosalvatori/ServiceBusExplorer/issues/607#issuecomment-2241635966

**Comment by maztan** [comment]
I will start a side project. Hope you don't mind. I just can't work with WinForms backed by business logic spaghetti. I know that it is the way things were done some time ago. I am not sure how far I will be able to go with this but I will start with no-gui version so that GUI can be added on top, be this Avalonia or some web GUI like Blazor. I am currently writing against .NET 8 so there will be cross-platform support. The hardest part I think is removing the custom classes that were added to this project, that can be replaced with the newly added ones from the new version of the SDK. The docs are suprisingly not that great if one wants to use all the functionality of the SDK and not just simple, "standard" tasks. I will probably have to read entire codebase of the SDK to be sure I am doing everything right and not repeat what already ia implemented. Again, I have only so much time but will do what I can. I suspect Microsoft will fill functionality gap soon anyway, so I treat this as a learning opportunity.

**Comment by SeanFeldman** [comment]
Don't want to discourage you with my comment but is it wise to do so w/o the UI? Cross-platform, modern UI, and the latest SDK are all interconnected. The tool needs a holistic redesign with a committed effort, which, no one can provide today. There are quite a few alternative tools that do less, but are modern.

I appreciate the effort if you still decide to go with your spike. There are features in the current implementation I'd not bother with. And I'd try to split between various service, w/o reuse-abuse. We know for sure that most of the usage is SB and EH. Perhaps starting with one of these (EH) as a POC and then deciding on the next steps

**Comment by SeanFeldman** [comment]
https://github.com/paolosalvatori/ServiceBusExplorer?tab=readme-ov-file#alternative-service-bus-management-tools

**Comment by maztan** [comment]
The fun fact here is that there is apparently no way to run any tests without active Azure subscription :)

**Comment by SeanFeldman** [comment]
When testing swimming skills, water is needed.

Fun fact indeed.

**Comment by maztan** [comment]
> When testing swimming skills, water is needed.
> 
> Fun fact indeed.

I would just like to test in my own swimming pool first but I can't have one as Microsoft keeps all the water to themselves.

**Comment by SeanFeldman** [comment]
You'd pay for your own swimming pool more than $10/m.

**Comment by maztan** [comment]
Ok, after spending a lot of time studying your codebase, I must say that it is written in a way I have hard time digesting. I am not a super programmer, that's just my experience, I am not in a position to judge. I have a question. Why wouldn't you shift your efforts towards alternatives that are already cross platform, like https://github.com/telstrapurple/PurpleExplorer that you mention on the website, for example? Why is Service Bus Explorer (this codebase) superior to the other one?

**Comment by sean-feldman-sh** [comment]
This project has been around for a very long time and provides a lot more than the other tools.
I'd suggest keeping this issue about the topic of AAD and open a separate issue to discuss unrelated topics.

**Comment by Danielovich** [comment]
Is it correct that I cannot, still, login with a managed identity (Entra/Azure AD) from the Service Bus Explorer ?

**Comment by GABRIELNGBTUC** [comment]
@Danielovich The current preview is able to use EntraId. So, I figure we should all try it out and report issues here until it makes it in the next stable release.

**Comment by SeanFeldman** [comment]
Support for Entra ID has been added. Closing.

**Comment by patelamit06** [comment]
<img width="1008" height="938" alt="Image" src="https://github.com/user-attachments/assets/6e931462-3390-4cdb-b333-d18da369b959" />
@ErikMogensen  Is there any forced policy to use internet explorer
